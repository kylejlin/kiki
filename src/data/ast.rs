#[derive(Clone, Debug)]
pub struct File {
    pub items: Vec<Item>,
}

#[derive(Clone, Debug)]
pub enum Item {
    Start(Ident),
    Struct(StructDef),
    Enum(EnumDef),
    Terminal(TerminalDef),
}

#[derive(Clone, Debug)]
pub struct StructDef {
    pub name: Ident,
    pub fieldset: Fieldset,
}

#[derive(Clone, Debug)]
pub struct EnumDef {
    pub name: Ident,
    pub variants: Vec<EnumVariant>,
}

#[derive(Clone, Debug)]
pub struct TerminalDef {
    pub name: Ident,
    pub variants: Vec<TerminalVariant>,
}

#[derive(Clone, Debug)]
pub enum Fieldset {
    Empty,
    Named(NamedFieldset),
    Tuple(TupleFieldset),
}

#[derive(Clone, Debug)]
pub struct NamedFieldset {
    pub fields: Vec<NamedField>,
}

#[derive(Clone, Debug)]
pub struct NamedField {
    pub name: IdentOrUnderscore,
    pub symbol: IdentOrTerminalIdent,
}

#[derive(Clone, Debug)]
pub struct TupleFieldset {
    pub fields: Vec<TupleField>,
}

#[derive(Clone, Debug)]
pub enum TupleField {
    Used(IdentOrTerminalIdent),
    Skipped(IdentOrTerminalIdent),
}

#[derive(Clone, Debug)]
pub struct EnumVariant {
    pub name: Ident,
    pub fieldset: Fieldset,
}

#[derive(Clone, Debug)]
pub struct TerminalVariant {
    pub name: TerminalIdent,
    pub type_: Type,
}

#[derive(Clone, Debug)]
pub enum Type {
    Unit,
    Path(Path),
    Complex(Box<ComplexType>),
}

#[derive(Clone, Debug)]
pub enum Path {
    One(Ident),
    Cons(Box<Path>, Ident),
}

#[derive(Clone, Debug)]
pub struct ComplexType {
    pub callee: Type,
    pub types: Vec<Type>,
}

pub use crate::data::cst::{Ident, IdentOrTerminalIdent, IdentOrUnderscore, TerminalIdent, Token};
