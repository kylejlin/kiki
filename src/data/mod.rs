pub mod ast;
pub mod cst;
pub mod machine;
pub mod table;
pub mod validated_file;

#[derive(Debug)]
pub enum KikiErr {
    Parse(ByteIndex, String, ByteIndex),
    NoStartSymbol,
    MultipleStartSymbols(Vec<ByteIndex>),
    NoTerminalEnum,
    MultipleTerminalEnums(Vec<ByteIndex>),
    SymbolsFirstLetteNotCapitalized(ByteIndex),
    DuplicateTerminalVariants(String, ByteIndex, ByteIndex),
    DuplicateNonterminals(String, ByteIndex, ByteIndex),
    UndefinedNonterminal(String, ByteIndex),
    UndefinedTerminal(DollarlessTerminalName, ByteIndex),
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub struct RustSrc(pub String);

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct ByteIndex(pub usize);

pub use crate::logic::oset::Oset;

use validated_file::DollarlessTerminalName;
