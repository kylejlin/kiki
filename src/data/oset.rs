#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Oset<T> {
    raw: Vec<T>,
}

impl<T> Oset<T> {
    pub fn new() -> Self {
        Self { raw: Vec::new() }
    }
}

impl<T> Oset<T>
where
    T: Ord,
{
    pub fn insert(&mut self, item: T) {
        match self.raw.binary_search(&item) {
            Ok(_) => {}
            Err(i) => self.raw.insert(i, item),
        }
    }

    pub fn contains(&self, item: &T) -> bool {
        self.raw.binary_search(item).is_ok()
    }
}

impl<T> FromIterator<T> for Oset<T>
where
    T: Ord,
{
    fn from_iter<I>(iter: I) -> Self
    where
        I: IntoIterator<Item = T>,
    {
        let mut raw: Vec<T> = iter.into_iter().collect();
        raw.sort();
        raw.dedup();
        Self { raw }
    }
}

impl<T> IntoIterator for Oset<T> {
    type Item = T;
    type IntoIter = std::vec::IntoIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        self.raw.into_iter()
    }
}
