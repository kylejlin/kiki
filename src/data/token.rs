pub use crate::data::*;

pub use crate::pipeline::parser::Token;

impl PartialEq for Token {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Token::Underscore(a), Token::Underscore(b)) => a == b,
            (Token::Ident(a), Token::Ident(b)) => a == b,
            (Token::TerminalIdent(a), Token::TerminalIdent(b)) => a == b,
            (Token::StartKw(a), Token::StartKw(b)) => a == b,
            (Token::StructKw(a), Token::StructKw(b)) => a == b,
            (Token::EnumKw(a), Token::EnumKw(b)) => a == b,
            (Token::TerminalKw(a), Token::TerminalKw(b)) => a == b,
            (Token::Colon(a), Token::Colon(b)) => a == b,
            (Token::DoubleColon(a), Token::DoubleColon(b)) => a == b,
            (Token::Comma(a), Token::Comma(b)) => a == b,
            (Token::LParen(a), Token::LParen(b)) => a == b,
            (Token::RParen(a), Token::RParen(b)) => a == b,
            (Token::LCurly(a), Token::LCurly(b)) => a == b,
            (Token::RCurly(a), Token::RCurly(b)) => a == b,
            (Token::LAngle(a), Token::LAngle(b)) => a == b,
            (Token::RAngle(a), Token::RAngle(b)) => a == b,
            _ => false,
        }
    }
}

impl Eq for Token {}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Ident {
    pub name: String,
    pub position: ByteIndex,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct TerminalIdent {
    pub name: DollarlessTerminalName,
    pub dollarless_position: ByteIndex,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Attribute {
    pub src: String,
    pub position: ByteIndex,
}
