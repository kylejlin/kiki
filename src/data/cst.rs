use crate::data::*;

#[derive(Clone, Debug)]
pub struct File {
    pub items: OptItems,
}

#[derive(Clone, Debug)]
pub enum OptItems {
    Nil,
    Cons(Box<OptItems>, FileItem),
}

#[derive(Clone, Debug)]
pub enum FileItem {
    Start(Ident),
    Struct(Struct),
    Enum(Enum),
    Terminal(TerminalEnum),
}

#[derive(Clone, Debug)]
pub struct Struct {
    pub name: Ident,
    pub fieldset: Fieldset,
}

#[derive(Clone, Debug)]
pub struct Enum {
    pub name: Ident,
    pub variants: OptEnumVariants,
}

#[derive(Clone, Debug)]
pub struct TerminalEnum {
    pub name: Ident,
    pub variants: OptTerminalEnumVariants,
}

#[derive(Clone, Debug)]
pub enum Fieldset {
    Empty,
    Named(NamedFieldset),
    Tuple(TupleFieldset),
}

#[derive(Clone, Debug)]
pub struct NamedFieldset {
    pub fields: NamedFields,
}

#[derive(Clone, Debug)]
pub enum NamedFields {
    One(NamedField),
    Cons(Box<NamedFields>, NamedField),
}

#[derive(Clone, Debug)]
pub struct NamedField {
    pub name: IdentOrUnderscore,
    pub symbol: IdentOrTerminalIdent,
}

#[derive(Clone, Debug)]
pub struct TupleFieldset {
    pub fields: TupleFields,
}

#[derive(Clone, Debug)]
pub enum TupleFields {
    One(TupleField),
    Cons(Box<TupleFields>, TupleField),
}

#[derive(Clone, Debug)]
pub enum TupleField {
    Used(IdentOrTerminalIdent),
    Skipped(IdentOrTerminalIdent),
}

#[derive(Clone, Debug)]
pub enum OptEnumVariants {
    Nil,
    Cons(Box<OptEnumVariants>, EnumVariant),
}

#[derive(Clone, Debug)]
pub struct EnumVariant {
    pub name: Ident,
    pub fieldset: Fieldset,
}

#[derive(Clone, Debug)]
pub enum OptTerminalEnumVariants {
    Nil,
    Cons(Box<OptTerminalEnumVariants>, TerminalEnumVariant),
}

#[derive(Clone, Debug)]
pub struct TerminalEnumVariant {
    pub name: TerminalIdent,
    pub type_: Type,
}

#[derive(Clone, Debug)]
pub enum Type {
    Unit,
    Path(Path),
    Complex(Box<ComplexType>),
}

#[derive(Clone, Debug)]
pub enum Path {
    One(Ident),
    Cons(Box<Path>, Ident),
}

#[derive(Clone, Debug)]
pub struct ComplexType {
    pub callee: Path,
    pub args: CommaSeparatedTypes,
}

#[derive(Clone, Debug)]
pub enum CommaSeparatedTypes {
    One(Type),
    Cons(Box<CommaSeparatedTypes>, Type),
}

#[derive(Clone, Debug)]
pub enum IdentOrUnderscore {
    Ident(Ident),
    Underscore,
}

#[derive(Clone, Debug)]
pub enum IdentOrTerminalIdent {
    Ident(Ident),
    Terminal(TerminalIdent),
}

#[derive(Clone, Debug)]
pub enum Token {
    Underscore,
    Ident(Ident),
    TerminalIdent(TerminalIdent),

    StartKw,
    StructKw,
    EnumKw,
    TerminalKw,

    Colon,
    DoubleColon,
    Comma,

    LParen,
    RParen,
    LCurly,
    RCurly,
    LAngle,
    RAngle,
}

#[derive(Clone, Debug)]
pub struct Ident {
    pub name: String,
    pub position: ByteIndex,
}

#[derive(Clone, Debug)]
pub struct TerminalIdent {
    pub name: DollarlessTerminalName,
    pub dollarless_position: ByteIndex,
}
