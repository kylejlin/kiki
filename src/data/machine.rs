use crate::data::{validated_file::DollarlessTerminalName, Oset};

#[derive(Debug, Clone)]
pub struct Machine {
    /// The first state is the start state.
    pub states: Vec<State>,
    pub transitions: Vec<Transition>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct State {
    pub items: Oset<Item>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Item {
    pub rule_index: RuleIndex,
    pub lookahead: Lookahead,
    /// The `dot` is the index of the symbol to the right of the dot.
    /// If the dot is at the end of the RHS, then `dot == right.len()`.
    pub dot: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum RuleIndex {
    Original(usize),
    Augmented,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Lookahead {
    Terminal(DollarlessTerminalName),
    Eof,
}

#[derive(Debug, Clone)]
pub struct Transition {
    pub from: StateIndex,
    pub to: StateIndex,
    pub symbol: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct StateIndex(pub usize);

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Symbol {
    Terminal(DollarlessTerminalName),
    Nonterminal(String),
}
