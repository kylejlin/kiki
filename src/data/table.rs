use crate::data::*;

#[derive(Debug, Clone)]
pub struct Table {
    pub dollarless_terminals: Vec<DollarlessTerminalName>,
    pub nonterminals: Vec<String>,
    pub actions: Vec<Action>,
    pub gotos: Vec<Goto>,
}

#[derive(Debug, Clone, Copy)]
pub enum Action {
    Shift(usize),
    Reduce(usize),
    Accept,
    Err,
}

#[derive(Debug, Clone, Copy)]
pub enum Goto {
    State(usize),
    Err,
}

#[derive(Debug, Clone, Copy)]
pub enum Quasiterminal<'a> {
    Terminal(&'a DollarlessTerminalName),
    Eof,
}

impl Table {
    pub fn state_count(&self) -> usize {
        self.actions.len() / (self.dollarless_terminals.len() + 1)
    }

    /// ## Panics
    /// 1. Panics if the terminal is not in the table.
    /// 2. Panics if the state is too large.
    pub fn action(&self, state: usize, terminal: Quasiterminal) -> Action {
        let i = self.action_index(state, terminal);
        self.actions[i]
    }

    /// ## Panics
    /// 1. Panics if the terminal is not in the table.
    /// 2. Panics if the state is too large.
    pub fn set_action(&mut self, state: usize, terminal: Quasiterminal, val: Action) {
        let i = self.action_index(state, terminal);
        self.actions[i] = val;
    }

    /// ## Panics
    /// 1. Panics if the terminal is not in the table.
    /// 2. Panics if the state is too large.
    fn action_index(&self, state: usize, quasiterminal: Quasiterminal) -> usize {
        let quasiterminal_index = match quasiterminal {
            Quasiterminal::Terminal(terminal) => self
                .dollarless_terminals
                .iter()
                .position(|t| t == terminal)
                .expect("Terminal not found in table"),
            Quasiterminal::Eof => self.dollarless_terminals.len(),
        };

        if state >= self.state_count() {
            let states = self.state_count();
            panic!("State {state} is too large. There are only {states} states.");
        }

        state * (self.dollarless_terminals.len() + 1) + quasiterminal_index
    }

    /// ## Panics
    /// 1. Panics if the nonterminal is not in the table.
    /// 2. Panics if the state is too large.
    pub fn goto(&self, state: usize, nonterminal: &str) -> Goto {
        let i = self.goto_index(state, nonterminal);
        self.gotos[i]
    }

    /// ## Panics
    /// 1. Panics if the nonterminal is not in the table.
    /// 2. Panics if the state is too large.
    pub fn set_goto(&mut self, state: usize, nonterminal: &str, val: Goto) {
        let i = self.goto_index(state, nonterminal);
        self.gotos[i] = val;
    }

    /// ## Panics
    /// 1. Panics if the nonterminal is not in the table.
    /// 2. Panics if the state is too large.
    fn goto_index(&self, state: usize, nonterminal: &str) -> usize {
        let nonterminal_index = self
            .nonterminals
            .iter()
            .position(|t| t == nonterminal)
            .expect("Nonterminal not found in table");

        if state >= self.state_count() {
            let states = self.state_count();
            panic!("State {state} is too large. There are only {states} states.");
        }

        state * self.nonterminals.len() + nonterminal_index
    }
}
