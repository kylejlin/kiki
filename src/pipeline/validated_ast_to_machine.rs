use crate::data::{
    ast::{Fieldset, NamedFieldset, TupleFieldset},
    machine::*,
    validated_file::*,
    KikiErr, Oset,
};
use std::collections::HashMap;
use std::collections::VecDeque;

/// Converts the AST to a finite state machine (FSM).
pub fn validated_ast_to_machine(file: &File) -> Result<Machine, KikiErr> {
    let builder = MachineBuilder::new(file);
    builder.build()
}

#[derive(Debug, Clone)]
struct MachineBuilder<'a> {
    context: ImmutContext<'a>,
    machine: Machine,
    queue: VecDeque<StateIndex>,
}

#[derive(Debug, Clone)]
struct ImmutContext<'a> {
    start_nonterminal_name: String,
    rules: Vec<Rule<'a>>,
    first_sets: HashMap<String, FirstSet>,
}

#[derive(Debug, Clone)]
struct FirstSet {
    terminals: Oset<DollarlessTerminalName>,
    contains_epsilon: bool,
}

#[derive(Debug, Clone)]
struct AugmentedFirstSet(Oset<Lookahead>);

impl MachineBuilder<'_> {
    fn new(file: &File) -> MachineBuilder {
        let context = ImmutContext::new(file);
        let start_state = context.get_start_state();
        MachineBuilder {
            context,
            machine: Machine {
                states: vec![start_state],
                transitions: Oset::new(),
            },
            queue: VecDeque::from([StateIndex(0)]),
        }
    }
}

impl ImmutContext<'_> {
    fn new(file: &File) -> ImmutContext {
        let rules: Vec<Rule> = file.get_rules().collect();
        let first_sets = get_first_sets(&rules);
        ImmutContext {
            start_nonterminal_name: file.start.clone(),
            rules,
            first_sets,
        }
    }
}

impl MachineBuilder<'_> {
    fn build(mut self) -> Result<Machine, KikiErr> {
        while let Some(state_index) = self.queue.pop_front() {
            self.enqueue_transition_targets(state_index);
        }
        Ok(self.machine)
    }

    fn enqueue_state_if_needed(&mut self, state: State) -> StateIndex {
        if let Some(index) = self.get_index_of_mergable(&state) {
            self.merge(index, state.items)
        } else {
            self.enqueue_new_state(state)
        }
    }

    fn get_index_of_mergable(&self, state: &State) -> Option<StateIndex> {
        self.machine
            .states
            .iter()
            .enumerate()
            .find_map(|(i, existing_state)| {
                if are_cores_equal(state, existing_state) {
                    Some(StateIndex(i))
                } else {
                    None
                }
            })
    }

    fn merge(&mut self, index: StateIndex, items: Oset<Item>) -> StateIndex {
        let were_items_added = self.add_items_if_needed(index, items);

        if were_items_added {
            self.queue.push_back(index);
        }

        return index;
    }

    /// Returns true if items were added.
    fn add_items_if_needed(&mut self, index: StateIndex, items: Oset<Item>) -> bool {
        let state = self.state_mut(index);
        let mut was_item_added = false;

        for item in items {
            if !state.items.contains(&item) {
                state.items.insert(item);
                was_item_added = true;
            }
        }

        was_item_added
    }

    fn enqueue_new_state(&mut self, state: State) -> StateIndex {
        let index = StateIndex(self.machine.states.len());
        self.machine.states.push(state);
        self.queue.push_back(index);
        index
    }

    fn enqueue_transition_targets(&mut self, state_index: StateIndex) {
        let next_symbols = self.get_symbols_right_of_dot(state_index);
        for symbol in &next_symbols {
            self.enqueue_transition_target(state_index, symbol);
        }
    }

    fn get_symbols_right_of_dot(&self, state_index: StateIndex) -> Oset<Symbol> {
        let state = self.state(state_index);
        state
            .items
            .iter()
            .filter_map(|item| self.get_symbol_right_of_dot(item))
            .collect()
    }

    fn get_symbol_right_of_dot(&self, item: &Item) -> Option<Symbol> {
        self.context.get_symbol_right_of_dot(item)
    }

    fn enqueue_transition_target(&mut self, state_index: StateIndex, symbol: &Symbol) {
        let target = self.get_transition_target(state_index, symbol);
        let target_index = self.enqueue_state_if_needed(target);
        let transition = Transition {
            from: state_index,
            to: target_index,
            symbol: symbol.clone(),
        };
        self.machine.transitions.insert(transition);
    }

    fn get_transition_target(&self, state_index: StateIndex, symbol: &Symbol) -> State {
        let items = self.get_transition_items(state_index, symbol);
        self.get_closure(&items)
    }

    fn get_closure(&self, items: &[Item]) -> State {
        self.context.get_closure(items)
    }

    fn get_transition_items(&self, state_index: StateIndex, symbol: &Symbol) -> Vec<Item> {
        let state = self.state(state_index);
        state
            .items
            .iter()
            .filter_map(|item| self.advance(item, symbol))
            .collect()
    }

    /// If `item` is `A -> alpha . B beta` and `symbol` is `B`,
    /// then this returns `Some(A -> alpha B . beta)`.
    fn advance(&self, item: &Item, symbol: &Symbol) -> Option<Item> {
        let right_of_dot = self.get_symbol_right_of_dot(item);
        if right_of_dot.as_ref() == Some(symbol) {
            Some(Item {
                rule_index: item.rule_index,
                lookahead: item.lookahead.clone(),
                dot: item.dot + 1,
            })
        } else {
            None
        }
    }
}

impl MachineBuilder<'_> {
    fn state(&self, index: StateIndex) -> &State {
        &self.machine.states[index.0]
    }

    fn state_mut(&mut self, index: StateIndex) -> &mut State {
        &mut self.machine.states[index.0]
    }
}

impl ImmutContext<'_> {
    fn get_start_state(&self) -> State {
        self.get_closure(&[Item {
            rule_index: RuleIndex::Augmented,
            lookahead: Lookahead::Eof,
            dot: 0,
        }])
    }

    fn get_closure(&self, items: &[Item]) -> State {
        let mut queue: VecDeque<Item> = items.iter().cloned().collect();
        let mut items = Oset::new();

        while let Some(next) = queue.pop_front() {
            if items.contains(&next) {
                continue;
            }

            self.enqueue_closure_implied_items(&mut queue, &next);
            items.insert(next);
        }

        State { items }
    }

    fn enqueue_closure_implied_items(&self, queue: &mut VecDeque<Item>, implicator: &Item) {
        for implied in self.get_closure_implied_items(implicator) {
            queue.push_back(implied);
        }
    }

    fn get_closure_implied_items(&self, item: &Item) -> Vec<Item> {
        match self.get_symbol_right_of_dot(item) {
            Some(Symbol::Nonterminal(name)) => {
                let lookaheads = self.get_augmented_first_after_dot(item);
                self.get_closure_implied_items_for_nonterminal(name, lookaheads)
            }
            Some(Symbol::Terminal(_)) | None => {
                vec![]
            }
        }
    }

    fn get_augmented_first_after_dot(&self, item: &Item) -> AugmentedFirstSet {
        let after_dot = self.get_symbol_sequence_after_dot(item);
        let first = self.get_first_of_symbol_sequence(after_dot);
        add_lookahead_if_needed(first, &item.lookahead)
    }

    fn get_symbol_sequence_after_dot<'a>(&'a self, item: &Item) -> Vec<Symbol> {
        match item.rule_index {
            RuleIndex::Original(rule_index) => {
                self.get_symbol_sequence_after_dot_for_original_rule(rule_index, item.dot)
            }
            RuleIndex::Augmented => self.get_symbol_sequence_after_dot_for_augmented_rule(item.dot),
        }
    }

    fn get_symbol_sequence_after_dot_for_original_rule(
        &self,
        rule_index: usize,
        dot: usize,
    ) -> Vec<Symbol> {
        let rule = &self.rules[rule_index];
        get_field_symbols_from_n_onwards(&rule.fieldset, dot)
    }

    fn get_symbol_sequence_after_dot_for_augmented_rule(&self, dot: usize) -> Vec<Symbol> {
        if dot == 0 {
            vec![Symbol::Nonterminal(self.start_nonterminal_name.clone())]
        } else {
            vec![]
        }
    }

    fn get_first_of_symbol_sequence(&self, symbols: impl IntoIterator<Item = Symbol>) -> FirstSet {
        let mut terminals: Oset<DollarlessTerminalName> = Oset::new();
        let mut contains_epsilon = true;

        for symbol in symbols {
            match symbol {
                Symbol::Terminal(name) => {
                    terminals.insert(name);
                    contains_epsilon = false;
                    break;
                }
                Symbol::Nonterminal(name) => {
                    let nonterminal_first_set = self.first_sets.get(&name).unwrap();
                    terminals.extend(nonterminal_first_set.terminals.iter().cloned());

                    if !nonterminal_first_set.contains_epsilon {
                        contains_epsilon = false;
                        break;
                    }
                }
            }
        }

        FirstSet {
            terminals,
            contains_epsilon,
        }
    }

    fn get_closure_implied_items_for_nonterminal(
        &self,
        nonterminal_name: String,
        lookaheads: AugmentedFirstSet,
    ) -> Vec<Item> {
        lookaheads
            .0
            .into_iter()
            .flat_map(|lookahead| {
                self.get_closure_implied_items_for_nonterminal_with_lookahead(
                    nonterminal_name.clone(),
                    lookahead,
                )
            })
            .collect()
    }

    fn get_closure_implied_items_for_nonterminal_with_lookahead(
        &self,
        nonterminal_name: String,
        lookahead: Lookahead,
    ) -> Vec<Item> {
        self.get_rule_indices_for_nonterminal(&nonterminal_name)
            .into_iter()
            .map(|rule_index| Item {
                rule_index: RuleIndex::Original(rule_index),
                lookahead: lookahead.clone(),
                dot: 0,
            })
            .collect()
    }

    fn get_rule_indices_for_nonterminal<'a>(
        &'a self,
        nonterminal_name: &'a str,
    ) -> impl Iterator<Item = usize> + 'a {
        self.rules
            .iter()
            .enumerate()
            .filter_map(move |(index, rule)| {
                if rule.constructor_name.type_name() == nonterminal_name {
                    Some(index)
                } else {
                    None
                }
            })
    }

    fn get_symbol_right_of_dot(&self, item: &Item) -> Option<Symbol> {
        match item.rule_index {
            RuleIndex::Original(rule_index) => {
                self.get_symbol_right_of_dot_for_original_rule(item.dot, rule_index)
            }
            RuleIndex::Augmented => self.get_symbol_right_of_dot_for_augmented_rule(item.dot),
        }
    }

    fn get_symbol_right_of_dot_for_augmented_rule(&self, dot: usize) -> Option<Symbol> {
        if dot == 0 {
            Some(Symbol::Nonterminal(self.start_nonterminal_name.clone()))
        } else {
            None
        }
    }

    fn get_symbol_right_of_dot_for_original_rule(
        &self,
        dot: usize,
        rule_index: usize,
    ) -> Option<Symbol> {
        let rule = &self.rules[rule_index];
        get_nth_field_symbol(dot, rule.fieldset)
    }
}

fn get_first_sets(rules: &[Rule]) -> HashMap<String, FirstSet> {
    let mut out = get_a_map_of_each_nonterminal_to_the_empty_set(rules);

    loop {
        let mut changed = false;

        todo!()
        
        if !changed {
            return out;
        }
    }
}

fn get_a_map_of_each_nonterminal_to_the_empty_set(rules: &[Rule]) -> HashMap<String, FirstSet> {
    let mut out = HashMap::new();
    for name in get_nonterminal_names(rules) {
        out.insert(
            name.to_owned(),
            FirstSet {
                terminals: Oset::new(),
                contains_epsilon: false,
            },
        );
    }
    out
}

fn get_nonterminal_names<'a>(rules: &'a [Rule<'a>]) -> Oset<&'a str> {
    rules
        .iter()
        .map(|rule| rule.constructor_name.type_name())
        .collect()
}

fn add_lookahead_if_needed(first: FirstSet, lookahead: &Lookahead) -> AugmentedFirstSet {
    if first.contains_epsilon {
        augment_with_lookahead(first, lookahead)
    } else {
        convert_first_set_to_augmented_as_is(first)
    }
}

fn augment_with_lookahead(first: FirstSet, lookahead: &Lookahead) -> AugmentedFirstSet {
    AugmentedFirstSet(
        first
            .terminals
            .into_iter()
            .map(Lookahead::Terminal)
            .chain(std::iter::once(lookahead.clone()))
            .collect(),
    )
}

fn convert_first_set_to_augmented_as_is(first: FirstSet) -> AugmentedFirstSet {
    AugmentedFirstSet(
        first
            .terminals
            .into_iter()
            .map(Lookahead::Terminal)
            .collect(),
    )
}

fn are_cores_equal(a: &State, b: &State) -> bool {
    is_core_subset(a, b) && is_core_subset(b, a)
}

fn is_core_subset(substate: &State, superstate: &State) -> bool {
    substate.items.iter().all(|sub| {
        superstate
            .items
            .iter()
            .any(|super_| sub.rule_index == super_.rule_index && sub.dot == super_.dot)
    })
}

fn get_nth_field_symbol(n: usize, fieldset: &Fieldset) -> Option<Symbol> {
    match fieldset {
        Fieldset::Empty => None,
        Fieldset::Named(named) => get_nth_field_symbol_from_named(n, named),
        Fieldset::Tuple(tuple) => get_nth_field_symbol_from_tuple(n, tuple),
    }
}

fn get_nth_field_symbol_from_named(n: usize, named: &NamedFieldset) -> Option<Symbol> {
    named.fields.get(n).map(|field| field.symbol.clone().into())
}

fn get_nth_field_symbol_from_tuple(n: usize, tuple: &TupleFieldset) -> Option<Symbol> {
    tuple
        .fields
        .get(n)
        .map(|field| field.symbol().clone().into())
}

fn get_field_symbols_from_n_onwards(fieldset: &Fieldset, n: usize) -> Vec<Symbol> {
    match fieldset {
        Fieldset::Empty => vec![],
        Fieldset::Named(named) => get_field_symbols_from_n_onwards_for_named(named, n),
        Fieldset::Tuple(tuple) => get_field_symbols_from_n_onwards_for_tuple(tuple, n),
    }
}

fn get_field_symbols_from_n_onwards_for_named(named: &NamedFieldset, n: usize) -> Vec<Symbol> {
    named
        .fields
        .iter()
        .skip(n)
        .map(|field| field.symbol.clone().into())
        .collect()
}

fn get_field_symbols_from_n_onwards_for_tuple(tuple: &TupleFieldset, n: usize) -> Vec<Symbol> {
    tuple
        .fields
        .iter()
        .skip(n)
        .map(|field| field.symbol().clone().into())
        .collect()
}
