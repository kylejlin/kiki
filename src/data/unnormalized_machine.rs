use crate::data::machine::{State, Transition};

use std::collections::HashSet;

#[derive(Debug, Clone)]
pub struct UnnormalizedMachine {
    /// The first state must be the start state.
    pub states: Vec<State>,
    pub transitions: HashSet<Transition>,
}
