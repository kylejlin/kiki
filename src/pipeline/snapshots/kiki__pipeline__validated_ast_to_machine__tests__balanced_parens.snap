---
source: src/pipeline/validated_ast_to_machine.rs
expression: machine
---
Machine {
    start: StateIndex(
        1,
    ),
    states: Oset {
        raw: [
            State {
                items: Oset {
                    raw: [
                        Item {
                            rule_index: Original(
                                0,
                            ),
                            lookahead: Terminal(
                                DollarlessTerminalName(
                                    "RParen",
                                ),
                            ),
                            dot: 0,
                        },
                        Item {
                            rule_index: Original(
                                1,
                            ),
                            lookahead: Terminal(
                                DollarlessTerminalName(
                                    "RParen",
                                ),
                            ),
                            dot: 0,
                        },
                        Item {
                            rule_index: Original(
                                1,
                            ),
                            lookahead: Terminal(
                                DollarlessTerminalName(
                                    "RParen",
                                ),
                            ),
                            dot: 1,
                        },
                        Item {
                            rule_index: Original(
                                1,
                            ),
                            lookahead: Eof,
                            dot: 1,
                        },
                    ],
                },
            },
            State {
                items: Oset {
                    raw: [
                        Item {
                            rule_index: Original(
                                0,
                            ),
                            lookahead: Eof,
                            dot: 0,
                        },
                        Item {
                            rule_index: Original(
                                1,
                            ),
                            lookahead: Eof,
                            dot: 0,
                        },
                        Item {
                            rule_index: Augmented,
                            lookahead: Eof,
                            dot: 0,
                        },
                    ],
                },
            },
            State {
                items: Oset {
                    raw: [
                        Item {
                            rule_index: Original(
                                1,
                            ),
                            lookahead: Terminal(
                                DollarlessTerminalName(
                                    "RParen",
                                ),
                            ),
                            dot: 2,
                        },
                        Item {
                            rule_index: Original(
                                1,
                            ),
                            lookahead: Eof,
                            dot: 2,
                        },
                    ],
                },
            },
            State {
                items: Oset {
                    raw: [
                        Item {
                            rule_index: Original(
                                1,
                            ),
                            lookahead: Terminal(
                                DollarlessTerminalName(
                                    "RParen",
                                ),
                            ),
                            dot: 3,
                        },
                        Item {
                            rule_index: Original(
                                1,
                            ),
                            lookahead: Eof,
                            dot: 3,
                        },
                    ],
                },
            },
            State {
                items: Oset {
                    raw: [
                        Item {
                            rule_index: Augmented,
                            lookahead: Eof,
                            dot: 1,
                        },
                    ],
                },
            },
        ],
    },
    transitions: Oset {
        raw: [
            Transition {
                from: StateIndex(
                    0,
                ),
                to: StateIndex(
                    0,
                ),
                symbol: Terminal(
                    DollarlessTerminalName(
                        "LParen",
                    ),
                ),
            },
            Transition {
                from: StateIndex(
                    0,
                ),
                to: StateIndex(
                    2,
                ),
                symbol: Nonterminal(
                    "Expr",
                ),
            },
            Transition {
                from: StateIndex(
                    1,
                ),
                to: StateIndex(
                    0,
                ),
                symbol: Terminal(
                    DollarlessTerminalName(
                        "LParen",
                    ),
                ),
            },
            Transition {
                from: StateIndex(
                    1,
                ),
                to: StateIndex(
                    4,
                ),
                symbol: Nonterminal(
                    "Expr",
                ),
            },
            Transition {
                from: StateIndex(
                    2,
                ),
                to: StateIndex(
                    3,
                ),
                symbol: Terminal(
                    DollarlessTerminalName(
                        "RParen",
                    ),
                ),
            },
        ],
    },
}
